### ISSUE TITLE
[Question] Where is documentation on how to link Google UMP consent results with Firebase consent?

### ISSUE THREAD

**Issue by batteredhedgehog** [firebase/firebase-unity-sdk#949]
### [REQUIRED] Please fill in the following fields:

  * Unity editor version: 2022.3.15f1
  * Firebase Unity SDK version: 11.7.0
  * Source you installed the SDK: FirebaseAnalytics (.unitypackage or Unity Package Manager)
  * Problematic Firebase Component: consent (Auth, Database, etc.)
  * Other Firebase Components in use: none (Auth, Database, etc.)
  * Additional SDKs you are using: Admob (Facebook, AdMob, etc.)
  * Platform you are using the Unity editor on: Linux (Mac, Windows, or Linux)
  * Platform you are targeting: Android / iOS (iOS, Android, and/or desktop)
  * Scripting Runtime: IL2CPP (Mono, and/or IL2CPP)
  * Pre-built SDK from the [website](https://firebase.google.com/download/unity) or open-source from this repo: _____

### [REQUIRED] Please describe the question here:
I am unable to find documentation on how to configure Firebase to update consent with data coming from google's UMP.
Please can you explain how to link the two, or point me to some documentation.

**Comment by paulinon** [comment]
Hi @batteredhedgehog,

You can use the `SetConsent` [function](https://firebase.google.com/docs/reference/unity/class/firebase/analytics/firebase-analytics#setconsent) to set the consent state. You may have to do some preliminary steps to [set up consent mode for your app](https://developers.google.com/tag-platform/security/guides/app-consent?consentmode=advanced&platform=android#update-consent) (depending on your platform). You may refer to this [example](https://github.com/firebase/firebase-unity-sdk/blob/main/analytics/testapp/Assets/Firebase/Sample/Analytics/UIHandler.cs#L127) in our quickstart to see how it's being used.

That said, I'll be closing this for now. Let me know if an issue arises so this ticket could be reopened.

**Comment by Remstam** [comment]
I think @batteredhedgehog wants some kind of conversion function from TCFStrings (which are created after UMP popup choices were made) to Firebase flags for _SetConsent_ function.

I am interested in this too, my current guess is to rely on [Tag Behaviour docs](https://developers.google.com/tag-platform/security/guides/implement-TCF-strings#tag_behavior). It is possible to take selected purposes string from shared preferences (Android) or NSUserDefaults (iOS) by _IABTCF_PurposeConsents_ key and go on from there.

But I am not sure if that is the right way to get values for _SetConsent_ call, wonder if there are any other opinions.

**Comment by batteredhedgehog** [comment]
@paulinon - thanks for the info.  I do understand how the SetConsent function is intended to work.  As @Remstam says, I'm looking for the correct way to extract consent information from TCF strings. 
Can you point me to a mapping between purposes and the consent mode v2 flags used in SetConsent ?
Should we be using PurposeConsents?  Do we need to look into google vendor consent purposes?
Any assistance would be very helpful.

**Comment by cometa93** [comment]
Bump

**Comment by OldSlash** [comment]
Any updates?

**Comment by llfabris** [comment]
up

**Comment by batteredhedgehog** [comment]
For anyone interested - I did the following:

- Delay initializing analytics and crashlytics until after UMP consent negotiation in my ApplicationManifest.xml using:
```
    <meta-data android:name="google_analytics_default_allow_analytics_storage" android:value="false" />
    <meta-data android:name="google_analytics_default_allow_ad_storage" android:value="false" />
    <meta-data android:name="google_analytics_default_allow_ad_user_data" android:value="false" />
    <meta-data android:name="google_analytics_default_allow_ad_personalization_signals" android:value="false" />
    <meta-data android:name="firebase_crashlytics_collection_enabled" android:value="false" />
```
- Once consent is negotiated I determine if google (IAB vendor 755) has been granted consent for purposes 1,3,4, and either consent of legitimate intereste for purpose 7.
- I enable crashlytics (using `Crashlytics.IsCrashlyticsCollectionEnabled = true` if either consent wasn't required (as determined by UMP) or if the consents in the previous step were given.
- I enable analytics using `FirebaseAnalytics.SetConsent(...)` if either consent wasn't required - or using the rules described [here](https://developers.google.com/tag-platform/security/guides/implement-TCF-strings#tcf_integration_behavior) based on the determined purpose consent status.

I dont really know if this is 100% correct - but its a start .  
Any guidance from google would be greatly appreciated.

**Comment by batteredhedgehog** [comment]
And here is some c# code snippets I wrote to help extract consent from the TCF string.
```
    private const string IABTCF_PurposeConsents = "IABTCF_PurposeConsents";
    private const string IABTCF_PurposeLegitimateInterests = "IABTCF_PurposeLegitimateInterests";
    private const string IABTCF_VendorConsents = "IABTCF_VendorConsents";
    private const string IABTCF_VendorLegitimateInterests = "IABTCF_VendorLegitimateInterests";
    private const string IABTCF_PublisherConsent = "IABTCF_PublisherConsent";
    private const string IABTCF_PublisherLegitimateInterests = "IABTCF_PublisherLegitimateInterests";

    public const int IABTCF_GOOGLE_VENDOR_ID = 755;

    private static bool IsBitSet(int nonZeroBasedBitIndex, string bitString)
    {
        return nonZeroBasedBitIndex > 0 && bitString != null && bitString.Length >= nonZeroBasedBitIndex && bitString[nonZeroBasedBitIndex - 1] == '1';
    }

    public static bool HasConsent(string purposeString, string vendorString, int purposeId, int? vendorId = null)
    {
        return IsBitSet(purposeId, ApplicationPreferences.GetString(purposeString)) && ((!vendorId.HasValue) || IsBitSet(vendorId.Value, ApplicationPreferences.GetString(vendorString)));
    }

    public static bool HasPurposeConsent(int purposeId, int? vendorId = null)
    {
        return HasConsent(IABTCF_PurposeConsents, IABTCF_VendorConsents, purposeId, vendorId);
    }

    public static bool HasLegitimateInterest(int purposeId, int? vendorId = null)
    {
        return HasConsent(IABTCF_PurposeLegitimateInterests, IABTCF_VendorLegitimateInterests, purposeId, vendorId);
    }

    public static bool HasPublisherConsent(int purposeId)
    {
        return HasConsent(IABTCF_PublisherConsent, null, purposeId);
    }

    public static bool HasPublisherLegitimateInterest(int purposeId)
    {
        return HasConsent(IABTCF_PublisherLegitimateInterests, null, purposeId);
    }
```

**Comment by binouze** [comment]
Hi, 

It seems that since [Firebase Android SDK 32.8.0](https://firebase.google.com/support/release-notes/android#2024-03-21) and [Firebase iOS SDK 10.23.0](https://firebase.google.com/support/release-notes/ios#version_10230_-_march_19_2024), Firebase Analytics gets a new feature:
 - Integrated parsing of [Transparency & Consent Framework (TCF)](https://developers.google.com/tag-platform/security/guides/implement-TCF-strings) data to retrieve user consent preferences.
 
Any ETA on when this will be available in the Unity SDK ?
Or maybe is it possible to just update the dependencies ?

**Comment by Remstam** [comment]
Ok, 11.9.0 release gets updates to Android 32.8.1 and iOS 10.24.0 respectively.
Does this mean it is no more necessary to manually call SetConsent as Analytics will now itself automatically get the values from TCF string internally?

**Comment by akshaymoonfrog** [comment]
Bump

**Comment by binouze** [comment]
Hi,
With the 11.9.0 update is it still necessary to modify the AndroidManifest.xml to add:
``` 
<meta-data android:name="google_analytics_default_allow_analytics_storage" android:value="false" />
<meta-data android:name="google_analytics_default_allow_ad_storage" android:value="false" />
<meta-data android:name="google_analytics_default_allow_ad_user_data" android:value="false" />
<meta-data android:name="google_analytics_default_allow_ad_personalization_signals" android:value="false" />
<meta-data android:name="firebase_crashlytics_collection_enabled" android:value="false" />
```
Or all this is done automatically based on the TCF values ?
How does this work for non UE users ? Will it reactivate it automatically or I need to do it manually ?

**Comment by ismetdegoo** [comment]
Why is the documentation so

**Comment by ismetdegoo** [comment]
Any updates?

**Comment by WBoda** [comment]
up

**Comment by argzdev** [comment]
Hi folks, could you kindly add a thumbs up emoji to the author's original post. This will help us keep track of issues under our radar. That said, I'll bring this up to our engineer sync to get some feedback. Thanks!

**Comment by stergios222** [comment]
up

**Comment by Iq51** [comment]
Bump

**Comment by nahrak** [comment]
Bump. Firebase shows a message, which reads "Ads personalization consent signals not detected".

This page describes how to enable consent mode with TCF (why false though?):
https://developers.google.com/tag-platform/security/guides/implement-TCF-strings#apps
<metadata android:name="google_analytics_tcf_data_enabled" android:value="false" />

Do we still need to manually set everything to false in AndroidManife4st.xml though?

**Comment by frolky** [comment]
Bump

**Comment by a-maurice** [comment]
Hey all,

Sorry for the lost wait in a response, with something like this we wanted to make sure we had the correct response, which took a while.

I believe that UMP needs to enable automatic TCF integration by setting the IABTCF_EnableAdvertiserConsentMode bit. This is partially documented on this page, including how to do so for Android and iOS apps: https://developers.google.com/tag-platform/security/guides/implement-TCF-strings#apps  By doing so, the Google Analytics SDK will be able to read TCF strings written by UMP, so you shouldn't need to call out all the additional flags, possibly just that one.

**Comment by batteredhedgehog** [comment]
Thanks for replying, but your answer is a bit vague.  Can you please provide us with a real implementation example,  remembering that this is a unity sdk question.

Lets say we enable the IABTCF_EnableAdvertiserConsentMode  in the AndroidManifest.xml (Android) and/or info.plist (iOS) bit as per the documentation you linked. 

Does this mean:

-  that in jurisdictions in which the UMP does not require consent, Firebase analytics will continue to function normally?
-  that in jurisdictions in which the UMP does require consent that NO information will be sent by the firebase analytics SDK before consent is obtained?
-  that in jurisdictions in which consent is required but not given, that we should go ahead and initialise analytics, and that analytics will respect this choice?
-  that when consent is given but subsequently withdrawn, analytics will respect this change?

**Comment by Remstam** [comment]
> Lets say we enable the IABTCF_EnableAdvertiserConsentMode in the AndroidManifest.xml (Android) and/or info.plist (iOS) bit as per the documentation you linked.

I guess this manifest/plist setup just allows Firebase not to take TCF strings into consideration:

- if _google analytics tcf data enabled_ is set to true (is this default value? Probably yes), then Firebase Analytics will rely on IAB flags
- if not enabled, then we have to use SetConsent method explicitly and dig out all the parameters for this method from IAB flags ourselves

As @a-maurice wrote IABTCF_EnableAdvertiserConsentMode bit has to be set also for everything to work out. Other CMP platforms state that they put that bit explicitly:

- Didomi: https://developers.didomi.io/cmp/mobile-sdk/google-consent-mode#enabling-consent-modes-tcf-integration
- OneTrust: https://my.onetrust.com/s/article/UUID-64f4f90c-38f8-bb69-e9b1-c8538dc9b774?language=en_US

Does UMP set this bit automatically? Probably no.

So the working scenario with UMP as for now looks like this:

1. Set _google analytics tcf data enabled_ = true in manifest/plist (maybe nothing to do here if true is default value)
2. UMP popup is being shown, user choose some settings
3. Set IABTCF_EnableAdvertiserConsentMode bit explicitly via SharedPreferences/NSUserDefaults
4. Init Firebase Analytics
5. Firebase Analyitcs reads TCF strings automatically and behaves appropriately

**Comment by stergios222** [comment]
the documentation at 
https://developers.google.com/tag-platform/security/guides/implement-TCF-strings#tcf_integration_behavior 
mentions the 
- ad_storage 
- ad_user_data 
- ad_personalization
being mapped from TCF Purposes, but not the analytics_storage one. Do we need to set this manually ?

**Comment by weakfl** [comment]
> So the working scenario with UMP as for now looks like this:
> 
> 1. Set _google analytics tcf data enabled_ = true in manifest/plist (maybe nothing to do here if true is default value)
> 2. UMP popup is being shown, user choose some settings
> 3. Set IABTCF_EnableAdvertiserConsentMode bit explicitly via SharedPreferences/NSUserDefaults
> 4. Init Firebase Analytics
> 5. Firebase Analyitcs reads TCF strings automatically and behaves appropriately

On iOS, this works to a certain degree only.

If you grant permissions and afterwards revoke them, `ad_storage`, `ad_user_data` and `ad_personalization` **are not reset** correctly, they are still set to `granted`.
Probably because the `UserDefaults.standard.string(forKey: "IABTCF_PurposeConsents")` is set to "0" and doesn't include all the required purposes.

So you definitely need to reset all permissions before showing the UMP dialog (again).

**Comment by ananttheant** [comment]
Guys, 

What should be the correct flow for when I only show the UMP after the login? 

In the `.plist` or `AndroidManifest`, I add them by default as granted and I need to make sure the Firebase is initialised at App start since I'd like to catch issues/crashes. I only show the UMP after the login is done and since the login comes at a later stage Is it fine to have those permissions as granted and `setConsent` after showing the UMP post login? (with an ability to further update consent on the settings panel)??? (Tagging @batteredhedgehog  since kind of had a similar situation)

**Comment by Remstam** [comment]
@a-maurice By the way, if _IABTCF_EnableAdvertiserConsentMode_ approach works and consent is transferred automatically, I would like to know what exactly the flag values are for custom analytics purposes. Could you please consider the idea of adding _GetConsent_ method? It should return current flags state, so we will know, what's been transferred after UMP choices are made.

**Comment by batteredhedgehog** [comment]
@a-maurice  - is there any update on how this really works?  the minimal information provided and limited documentation isn't sufficient for us devs to implement firebase analytics with UMP with any confidence that we're satisfying privacy requirements.

**Comment by ynnob** [comment]
i also wonder how all this works together iwth opt-in option the docs explain.
I would lie to initially have Firebase Disabled and then show the UMP to the user. But does that enable my firebase stuff or does this only handle Google Mobile Ad stuff. I am really confused by the amount of different documentations.

Opt-In example:
```
        <!-- Disable Crashalytics as default -->
        <meta-data
            android:name="firebase_crashlytics_collection_enabled"
            android:value="false" />

        <!-- Disable Firebase Analytics -->
        <meta-data
            android:name="firebase_analytics_collection_enabled"
            android:value="false" />

        <!-- Disable Performance Monitor -->
        <meta-data
            android:name="firebase_performance_collection_enabled"
            android:value="false" />
```

Then manually i would call:
`FirebaseCrashlytics.getInstance().setCrashlyticsCollectionEnabled(true);`

@ananttheant did you figure out what to do? I think i aim for the same setup. I kinda hate the idea to show multiple popups on first start to get the user consent. One Consent request with UMP would be nice.

**Comment by ananttheant** [comment]
@ynnob 

I used [this](https://developers.google.com/tag-platform/security/guides/app-consent?consentmode=basic&platform=ios#update-consent) documentation.

By default, I've set the consent as `granted`, and used the `eu_consent_policy` set as false so that I can initialise the Firebase normally without any change in its flow. After the login, we show the CMP popup, overriding the value for consents. 

Then finally you can pass on the consent values to the respective 3rd party SDKs

**Comment by emfigura** [comment]
Hey @a-maurice - do you have any guidance on how to properly do this? Or if anyone is working on a doc?

**Comment by ynnob** [comment]
i invested a bit more time and decided to split AdMob consent from Firebase because they don't really seem to be connected as thight as i expected them to be.

Also as far as i understood Settings set by UMP will (if a google service uses them) always be more important then the settings in the interface (e.g AdMob default personalized ads / non personalized ads will be overwritten through the UMP Consent selection)

So i opted to seperatly let the user opt-in and out of firebase while also doing the same with AdMbo and UMP.

That way i know i did everything to inform the user about the services i use, let them opt-in and out and also use the UMP if firebase at soem point accesses these setting.

**Comment by batteredhedgehog** [comment]
I think @ynnob is probably correct - its **safest** to ask for each explicitly - but this is gives a very poor user experience: The app might then have  IDFA explainer, ATT popup, UMP consent popup, and finally analytics consent popup...  If that happened to me when I installed a game, I'd uninstall it immediately.  As it is I have the first three - which is still too many.

On a side note, Using the configuration i mention above - i.e. sets consent mode flags to false and update flags once consent is given - the first_open event is never posted - meaning that the New Users count in Google Analytics is always 0.  If anyone can think of a way around when users do give consent - I'm all ears.

**Comment by ananttheant** [comment]
Well, the easiest way is to show the CMP popup first (before ATT), if the user wants to change the consent for specific things they can do it using manage consent option. (If you see it properly, it has multiple purposes) 

In the end, once the consent popup is done, you can fetch all the purposes and pass it on to your ads SDK and firebase. 

By checking if the user has declined the use of saving data related to ads you can also avoid showing ATT popup (this was mentioned by Google itself if their QnA)

**Comment by ananttheant** [comment]
Hey @paulinon / @a-maurice

I have a question regarding this statement 

> In accordance with the updates to consent mode for traffic in European Economic Area (EEA), a value of `eu_consent_policy` can be set for `ad_user_data` and `ad_personalization`, meaning consent is denied only for users in regions subject to the EU User Consent Policy.

As mentioned #[here](https://developers.google.com/tag-platform/security/guides/app-consent?consentmode=advanced&platform=ios#default-consent)

This is a bit confusing and I need to understand what exactly this looks like with example:
1. Does this mean that we can
```
<key>eu_consent_policy</key> <false/>
<key>GOOGLE_ANALYTICS_DEFAULT_ALLOW_ANALYTICS_STORAGE</key> <true/>
<key>GOOGLE_ANALYTICS_DEFAULT_ALLOW_AD_STORAGE</key> <true/>
<key>GOOGLE_ANALYTICS_DEFAULT_ALLOW_AD_USER_DATA</key> <true/>
<key>GOOGLE_ANALYTICS_DEFAULT_ALLOW_AD_PERSONALIZATION_SIGNALS</key> <true/>
```

2. Or does this mean
```
<key>GOOGLE_ANALYTICS_DEFAULT_ALLOW_ANALYTICS_STORAGE</key> <true/>
<key>GOOGLE_ANALYTICS_DEFAULT_ALLOW_AD_STORAGE</key> <true/>
<key>GOOGLE_ANALYTICS_DEFAULT_ALLOW_AD_USER_DATA</key> <string>eu_consent_policy</string>
<key>GOOGLE_ANALYTICS_DEFAULT_ALLOW_AD_PERSONALIZATION_SIGNALS</key> <string>eu_consent_policy</string>
```

3. Or does this mean
```
<key> ad_user_data</key> <string>eu_consent_policy</string>
<key> ad_personalization</key> <string>eu_consent_policy</string>
<key>GOOGLE_ANALYTICS_DEFAULT_ALLOW_ANALYTICS_STORAGE</key> <true/>
<key>GOOGLE_ANALYTICS_DEFAULT_ALLOW_AD_STORAGE</key> <true/>
<key>GOOGLE_ANALYTICS_DEFAULT_ALLOW_AD_USER_DATA</key> <true/>
<key>GOOGLE_ANALYTICS_DEFAULT_ALLOW_AD_PERSONALIZATION_SIGNALS</key> <true/>
```

**Comment by plthatfunkymusic** [comment]
Hi @ananttheant  - I think it's (2) in your list above as looking at the logs this looks like a valid setting.  What I'd really love to know is how we read the value for a given user. i.e. how to determine if they fall under the EU jurisdiction so that we can correctly set the default consent in our App, and only request updated consent from users in the EU (default I guess is DENIED, and then ask them if they are happy for it to be GRANTED).  I was hoping this would be possible for all 4 of the settings (I'd default them all to eu_consent_policy) and not bother users outside of the EU.  

I'm not currently interested in advertising (I don't serve Ads), but do want to collect analytics data.  It would be a shame to lose lots of analytics data because people believe they are being tracked across services, when really it's for debugging and improved features within the App itself.

What's not clear to me, is what are the current default values, as I currently don't set any of these and yet I'm still getting analytics data.  Is this something different - is this analytics used by Google themselves, or by 3rd party providers?  And, it's not clear why any of this is needed as the App already declares analytics are collected as required by Google policy.

In short, it would be good to know how to read the current (default) consent status....

**Comment by alkanyunus** [comment]
Hi,
I want to share my works shortly what I've done.
I use [Google UMP](https://developers.google.com/admob/unity/privacy) but I prefer to use it in native sides both objective-c and java because of my custom sdk. I extract TCF string purpose for `CanShowPersonalizedAds` and `CanTrackAnalytics` so that I can use them in my unity c# side.

1- if user is in EEA, GDPR compliance must be done before initializing any analytics or data collection systems.(in this case FirebaseAnalytics)
2- Get consent status from Google UMP at app startup.
3- Check dependency of Firebase
4- Set consent of FirebaseAnalytics

My code example below:

```csharp
 
if (!string.IsNullOrEmpty(UserId))
{
    FirebaseAnalytics.SetUserId(UserId);
    Crashlytics.SetUserId(UserId);
}

// allow ad personalization signal
var adPersonalizationSignalsStatus = CmpService.Instance.CanShowPersonalizedAds ? "true" : "false";
FirebaseAnalytics.SetUserProperty(FirebaseAnalytics.UserPropertyAllowAdPersonalizationSignals, adPersonalizationSignalsStatus);

// firebase consent
var canTrackAnalytics = CmpService.Instance.CanTrackAnalytics;
var consentStatus = canTrackAnalytics ? ConsentStatus.Granted : ConsentStatus.Denied;
FirebaseAnalytics.SetConsent(new Dictionary<ConsentType, ConsentStatus>
{
    { ConsentType.AdPersonalization, consentStatus },
    { ConsentType.AdStorage, consentStatus },
    { ConsentType.AdUserData, consentStatus },
    { ConsentType.AnalyticsStorage, consentStatus },
});

FirebaseAnalytics.SetAnalyticsCollectionEnabled(true);
```

`CmpService` is a wrapper class of Google UMP library, and this code is called after firebase dependency checked in case of `DependencyStatus.Available`

Unity PostProcess scripts to inject necessary fields with default values

`iOS`
```csharp
private const string KeyGoogleAnalyticsDefaultAllowAnalyticsStorage = "GOOGLE_ANALYTICS_DEFAULT_ALLOW_ANALYTICS_STORAGE";
private const string KeyGoogleAnalyticsDefaultAllowAdStorage = "GOOGLE_ANALYTICS_DEFAULT_ALLOW_AD_STORAGE";
private const string KeyGoogleAnalyticsDefaultAllowAdUserData = "GOOGLE_ANALYTICS_DEFAULT_ALLOW_AD_USER_DATA";
private const string KeyGoogleAnalyticsDefaultAllowPersonalizationSignals =
    "GOOGLE_ANALYTICS_DEFAULT_ALLOW_AD_PERSONALIZATION_SIGNALS";

private void EditInfoPlist(string path)
{
    var plistPath = path + "/Info.plist";
    var plist = new PlistDocument();
    plist.ReadFromString(File.ReadAllText(plistPath));

    // ... other processing
    AddFirebaseAnalyticsKeys(plist);

    File.WriteAllText(plistPath, plist.WriteToString());
}

private static void AddFirebaseAnalyticsKeys(PlistDocument plist)
{
    // Consent Mode v2
    plist.root.SetBoolean(KeyGoogleAnalyticsDefaultAllowAnalyticsStorage, true);
    plist.root.SetBoolean(KeyGoogleAnalyticsDefaultAllowAdStorage, true);
    plist.root.SetBoolean(KeyGoogleAnalyticsDefaultAllowAdUserData, true);
    plist.root.SetBoolean(KeyGoogleAnalyticsDefaultAllowPersonalizationSignals, true);
    
    UnicornLogger.Log("Firebase Analytics keys are injected to the plist file.", LogPrefix);
}
```

`Android`
```csharp
private const string TagMetaData = "meta-data";
private const string KeyMetaDataGoogleAnalyticsDefaultAllowAnalyticsStorage = "google_analytics_default_allow_analytics_storage";
private const string KeyMetaDataGoogleAnalyticsDefaultAllowAdStorage = "google_analytics_default_allow_ad_storage";
private const string KeyMetaDataGoogleAnalyticsDefaultAllowAdUserData = "google_analytics_default_allow_ad_user_data";
private const string KeyMetaDataGoogleAnalyticsDefaultAllowPersonalizationSignals =
    "google_analytics_default_allow_ad_personalization_signals";   

private static void ProcessAndroidManifest(string path)
{
    var manifestPath = Path.Combine(path, "src/main/AndroidManifest.xml");
    XDocument manifest;
    try
    {
        manifest = XDocument.Load(manifestPath);
    }
    #pragma warning disable 0168
    catch (IOException exception)
        #pragma warning restore 0168
    {
        UnicornLogger.LogWarning("AndroidManifest.xml is missing.", LogPrefix);
        return;
    }

    // Get the `manifest` element.
    var elementManifest = manifest.Element("manifest");
    if (elementManifest == null)
    {
        UnicornLogger.LogWarning("AndroidManifest.xml is invalid.", LogPrefix);
        return;
    }

    var elementApplication = elementManifest.Element("application");
    if (elementApplication == null)
    {
        UnicornLogger.LogWarning("AndroidManifest.xml is invalid.", LogPrefix);
        return;
    }

    // ... other processing

    AddFirebaseAnalyticsKeys(elementApplication);
    
    // Save the updated manifest file.
    manifest.Save(manifestPath);
}

private static void AddFirebaseAnalyticsKeys(XElement elementApplication)
{
    // Consent Mode v2
    elementApplication.Add(CreateElement(TagMetaData, KeyMetaDataGoogleAnalyticsDefaultAllowAnalyticsStorage, "true"));
    elementApplication.Add(CreateElement(TagMetaData, KeyMetaDataGoogleAnalyticsDefaultAllowAdStorage, "true"));
    elementApplication.Add(CreateElement(TagMetaData, KeyMetaDataGoogleAnalyticsDefaultAllowAdUserData, "true"));
    elementApplication.Add(CreateElement(TagMetaData, KeyMetaDataGoogleAnalyticsDefaultAllowPersonalizationSignals, "true"));
    
    UnicornLogger.Log("Firebase Analytics keys are injected to android manifest file.", LogPrefix);
}

/// <summary>
/// Creates and returns a <c>tag</c> element with the given name and value.
/// </summary>
private static XElement CreateElement(string tag, string name, object value)
{
    var tagElement = new XElement(tag);
    tagElement.Add(new XAttribute(AndroidNamespace + "name", name));

    if (value != null)
    {
        tagElement.Add(new XAttribute(AndroidNamespace + "value", value));
    }

    return tagElement;
}
```

Post processing scripts look having missing parts but you'll understand what it meant to be.

Feel free to ask and we should keep continue pushing this issue :) I believe there is some missing parts in here.

**Comment by DiovanM** [comment]
@alkanyunus hey thank you very much for sharing your solution, as far as I can see it's the best we can do with the information we have at the moment. Would you mind sharing a snippet of how you extract the TCF string purpose? I'm having difficult understanding which TCF string would match AdStorage and AnalyticsStorage consent types

**Comment by Remstam** [comment]
A few days ago I ran into this interesting page:
https://developers.google.com/funding-choices/fc-sdks

So it looks like things became a little bit simplified. If User Messaging Platform is used as CMP, it seems we at least can go with this:

- Show UMP popup, make the choice
- Read _UMP_consentModeValues_ from SharedPreferences/NSUserDefaults, parse values for all 4 consent modes
- Convert and put them into Firebase Analytics SetConsent method

Still not sure if automatic consent collection somehow works, but it looks like _IABTCF_EnableAdvertiserConsentMode_ setup may not be needed anymore. To make automatic collection work fine, Firebase has to be inited before UMP popup shows on. The negative effect on that is if Firebase fails to init for some reason, you can go eat rocks for a while.

As for now I go with manual route, but it also has it's flaw - during the (probably very short) period of time where Firebase is already inited, but not called it's SetConsent (UMP popup is coming up yet), all the LogEvent and ILRD calls may lose data due to default denied restrictions.

**Comment by batteredhedgehog** [comment]
> [@batteredhedgehog](https://github.com/batteredhedgehog) I took this approach and I do see the first_open event fired immediately after consent is given (checked w/ Charles + GA4), so I presume this has been fixed (Firebase Unity SDK 13.3.0).

@Nomad1108 - In Google Analytics I still have 0 for new users on android, but  I do get new user stats for IOS. In both platforms I'm using use the same consent setup, initialisation sequencing and code.

**Comment by batteredhedgehog** [comment]
> [@batteredhedgehog](https://github.com/batteredhedgehog) I build on Android, my flow is:
> 
> 1 - Disable Analytics and Crashlytics collection via Manifest.xml
> 2 - Set the default consent states to false via Manifest.xml
> 3 - Update the consent state based on the user interaction with Google UMP or custom popup 
> 4 - Set custom User Properties (optional) `FirebaseAnalytics.SetUserProperty(...)`
> 5 - Enable Analytics and Crashlytics collection

I have a similar sequence, but I wait for a known consent status before i initiate firebase analytics. 
When the game is first opened, the consent status is unknown, so firebase analytics initialisation is delayed until after the UMP ConsentInformation.Update(...) call.   When I wrote the implementation somewhere in the documentation this was recommended - although it no longer says this.   The current android documentation ( https://developers.google.com/admob/android/privacy/consent-mode ) seems to suggest that we should initialize firebase analytics first.
This _may_ be why I'm not seeing the first_open events on android (although I'm still confused as to why I see them on IOS)
